Sources/Utility/ProgressAnimation.swift

```diff
@@ -0,0 +1,290 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Basic
+
+/// A protocol to operate on terminal based progress animations.
+public protocol ProgressAnimationProtocol {
+    func update(progress: Int, total: Int, text: String)
+    func complete(success: Bool)
+    func clear()
+}
+
+/// A single line ninja-like progress animation.
+public final class SingleLineNinjaProgressAnimation: ProgressAnimationProtocol {
+    private struct State: Hashable {
+        let progress: Int
+        let total: Int
+    }
+
+    private let stream: OutputByteStream
+    private var displayedStates: Set<State> = []
+
+    init(stream: OutputByteStream) {
+        self.stream = stream
+    }
+
+    public func update(progress: Int, total: Int, text: String) {
+        let state = State(progress: progress, total: total)
+        guard !displayedStates.contains(state) else { return }
```

Comment:
I don't get it, why do we need the state?

Reply:
This was adapted from `SingleLinePercentProgressAnimation` which uses a `Set<Int>` to avoid repeating steps. We could get rid of it in both cases: it doesn't provide enormous value IMHO.